#1223 feature: sorting key maps

maksimowiczm opened the pull request:
I've made some progress on the sorting feature.

I've created a simple UI where users can choose the field priority and sort order of the fields, which are applied to the home key maps list. I don't think a bottom sheet is a good idea for this especially with the drag interactions, but I couldn't think of a better alternative. (I replaced help button 😅)

Sorting preferences are stored in `PreferenceRepository`. I think it would be annoying to have to apply sorting every time the app is launched.

I'm unsure about using sorting in `ListKeyMapsUseCaseImpl`. Currently, key maps are sorted every time, which might result in a poor user experience if someone has a large number of key maps. I think it might be a good idea to create a separate screen for searching, where key maps could be sorted.

It might also be a good idea to implement sorting with SQL for better performance.

Additionally, there doesn't seem to be an obvious way to sort triggers and options, so I need some help with that.

Closes #1223

GfEW commented:
I can't currently test PRs, but I'm not really into coding, anyway. 
I can, however, comment on the UI/UX side of things. Some remarks:

1. persistency
> Sorting preferences are stored

Makes sense, sort/filter prefs should indeed persist until changed or reset by the user. To keep forgetful users from missing something, it's important to clearly indicate whether the current listing is filtered.

2. sort delay, main vs. separate list
> Currently, key maps are sorted every time, which might result in a poor user experience if someone has a large number of key maps. I think it might be a good idea to create a separate screen for searching, where key maps could be sorted.

Is this really an issue? Could you give some figures of the expected sort delays with 30, 100 and 300 mappings, or estimates thereof?

Sorting and filtering can be _much_ more useful in the main listing than only in a separate search screen with limited functionality. To justify the inferior, latter option, the performance _difference_ would have to exceed 1.5 s with 30 mappings, or 3.5 s with 100 mappings (which already is a lot, but mainly affects the relatively small fraction of users who know they are stretching the limits).

I think it's reasonable to assume only very few will end up using more than a few hundreds mappings. And if done right, sorting a list of a few hundreds items shouldn't take more than one to two seconds, even on ageing hardware (for a relevant example, see the apps list in [AppManager](https://github.com/MuntashirAkon/AppManager)).

3. sorting of triggers or options
> there doesn't seem to be an obvious way to sort triggers and options

Ideally, sorting by triggers would be consistent with their exact name strings as they are presented to the user (including Left/Right, parentheses etc.), e. g. 
* "Shift Right (Any device * Do not remap) + Delete"

Sorting by options is trivial for mappings that have either no or exactly one option enabled. (As with every key, "no option enabled" is the smallest possible value (empty string), coming before all possibly enabled options.)

The case of multiple options being enabled is a bit awkward, because their presentation order currently doesn't follow an obvious scheme. However, sorting according to the presented string would be the best we could do here, too.

p. s. Nice PR ID.

p. p. s. I've improved the _sorted list_ example in #1223, to make it visually easier to comprehend.

sds100 commented:
First I'm going to complete version 2.7 and then I will work on this.

sds100 commented on app/src/main/java/io/github/sds100/keymapper/constraints/ConstraintState.kt:
```diff
@@ -11,3 +12,41 @@ data class ConstraintState(
+            return 0
+        }
+
+        for (i in 0 until keyMap.constraintState.constraints.size.coerceAtMost(otherKeyMap.constraintState.constraints.size)) {
```
Can you put this into a separate variable and maybe explain what coerceAtMost is doing?

sds100 commented on app/src/main/java/io/github/sds100/keymapper/mappings/keymaps/KeyMapField.kt:
```diff
@@ -0,0 +1,19 @@
+import io.github.sds100.keymapper.constraints.KeyMapConstraintsComparator
+import io.github.sds100.keymapper.mappings.keymaps.trigger.KeyMapTriggerComparator
+
+enum class KeyMapField {
```
I would move this to the `sorting` package and rename it SortField? I could see one adding more ways to sort that aren't strictly key map fields?

maksimowiczm commented:
I worked on it a bit.

---

@GfEW

> 2. sort delay, main vs. separate list

I might have underestimated the performance of modern devices. I don't think it will be an issue with such a "small" list.

---

I applied your suggestions. @sds100

---

Current sort order for each field:

- **Trigger**
1. Keys are sorted one by one until one trigger runs out of keys (different size of keys list)

    Keys are sorted by
    1. Code
    3. Click type (short, long, double)
    4. Device
    5. "Do not remap" option
2. Keys' length - the shorter the better
3. Trigger mode (sequence, parallel)

- **Constraints**
Sorted by their name. If names are identical, they are sorted by their value (e.g., package name, Bluetooth address, etc.).

- **Actions** 
Sorted by their name (e.g., all `Launch app` actions are treated as the same name). I am currently working on sorting them by their value, similar to constraints.

- **Options** are not sorted at all. Not sure where to find all options in the code I need some help with that. @sds100

---

I think users should be able to choose sorting options and arrange them according to their needs. For now, I’ve implemented a very basic UI with all field.


https://github.com/user-attachments/assets/fff98bf5-8e76-4fcf-b9e7-a3683faff7b2

---

### What's next?

- I will finish sorting actions by their value
- I will try to sort options
- Think about user interface

maksimowiczm commented:
So, I did some research and I came up with a solution where the sort fields are displayed as chips in a scrollable row. It could also wrap, but at this width, it looks awkward because one element will jump between the first and second row.

The bottom sheet could also be used with filters, where it would be added below the sorting feature.

This is a similar solution to the one used by TikTok.

<img src="https://github.com/user-attachments/assets/1d925de5-18ff-4717-bdf9-9b79a3c0d39e" width=400/>

sds100 commented:
Hey! I think the chip row option is good, as you mention, it provides more space to add extra options. Can you merge the `develop` branch into this so the merge conflicts are solved?

sds100 commented on app/src/main/java/io/github/sds100/keymapper/mappings/keymaps/trigger/KeyMapTrigger.kt:
```diff
@@ -165,3 +166,54 @@ object KeymapTriggerEntityMapper {
+
+// Comparator.reversed() requires API level 24
+class KeyMapTriggerComparator(
+    private val reverse: Boolean = false,
```
There should be a Kotlin way to do reverse comparators instead of having to add it manually. Is this useful? https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.comparisons/reverse-order.html

maksimowiczm replied:
I tried to find a Kotlin way and only suitable method was https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.comparisons/reversed.html but it requires at least android 7

sds100 commented on app/src/main/java/io/github/sds100/keymapper/mappings/keymaps/trigger/KeyMapTrigger.kt:
```diff
@@ -165,3 +166,54 @@ object KeymapTriggerEntityMapper {
 }
+
+// Comparator.reversed() requires API level 24
+class KeyMapTriggerComparator(
```
Should this not be called KeyMapComparator? Can you put it inside the KeyMap file?

maksimowiczm replied:
Yeah name and location is a bit off. There are comparators  for every `SortField` and they are used for sorting `KeyMap` by given field.

`KeyMapTriggerComparator`, `KeyMapActionsComparator`, `KeyMapConstraintsComparator`, `KeyMapOptionsComparator` 
All of them are located kinda randomly.

sds100 replied:
If there isn't a good correspondence with the data classes then move the comparators to individual files in side a `sorting/comparators` package?

GfEW commented:
Sorry for the long delay - very busy days here. 

> sort fields are displayed as chips in a scrollable row

Chips have proven to serve well for simple sort operations, and the established UI paradigm helps users to get along easily. 

When sorting by multiple fields, I don't fully understand your presentation, yet. The tinted chips are "active", the grey one "inactive", that obviously works. But can the chips themselves be sorted? If not, how else do you indicate sort priorities, i. e. which active field matters most, which second, etc.? (In this regard, your "[very basic UI](https://github.com/keymapperorg/KeyMapper/pull/1337#issuecomment-2481370751)" was appreciably clear.)
 
> The bottom sheet could also be used with filters, where it would be added below the sorting feature.

Sounds promising! Great to see progress in these improvements to keymapper's long awaiting list UI.

sds100 commented:
I guess if its critical to have the ability to also sort the filters by priority, then the first implementation with a drag-drop list is the best option? But is it not overkill? Surely for 99% of users it is sufficient to sort by just trigger or the actions? To give that extra level of organization the grouping/folder feature should be implemented #320

maksimowiczm commented:
Chips can now be moved around. I don't have much experience implementing lists with draggable elements, so it looks a bit clumsy. I went a multi-line chips list because IMO it's more convenient for moving them.

I'm not sure whether animations improve the user experience or not.

This is without animations.

[Screen_recording_20241219_173226.webm](https://github.com/user-attachments/assets/c8d7eacd-a14e-48cc-8368-302b600d2ad5)

This is with animations. There are some issues with shadow if the user moves item too quickly

[anim.webm](https://github.com/user-attachments/assets/2e717960-8246-48a8-9d8a-24fa7468e0c3)

sds100 commented:
Thanks for the demo. I think if we want to reorder them then the list (not chips) provides a better UX and makes it clearer it is possible to reorder them cos its similar to the trigger list. What does @GfEW think?

GfEW commented:
> Thanks for the demo. I think if we want to reorder them then the list (not chips) provides a better UX and makes it clearer it is possible to reorder them cos its similar to the trigger list. What does @GfEW think?

What do I think, mhh ;-) -  All three variants have their points. My 2 cents:

#### (a) sortable list
* `-` not a common android UI paradigm
* `+` corresponds to the vertical nature of the main list
* `+` most cultures easily understand the topmost item as dominant / most important (cf. "_top_ priority")
* `+` mostly self-explanatory, therefore little need to explicitely label priorities (especially given the above-average technical understanding of keymapper userbase)

#### (b) sortable chips
* `-` not a common android UI paradigm
* `-` the horizontal alignement where "the leftmost chip matters most" isn't quite as evident (let alone in RTL languages)
* `-` due to the weaker sense of being ordered, users might also suppose that "the chip tapped last matters most", and without testing, it's not obvious which of the two sort principles holds
* `*` would profit more than (a) from explicit prio labels (possibly at the cost of visual cleanliness)
* `+` once understood, (b) is as powerful as (a) and works as well

#### (c) fixed chips (simple sort)
* `+` by far the most common approach of these three
* `+` very easy to understand, even by your typical grandma
* `-` neither of the above points poses as much of a merit in a key mapping app as they do in apps addressing a more general audience
* `-` lacks the fine grained control offered by (a) and (b) (e. g. to sub-sort same mappings by devices)

Personally, I'm slightly inclined towards (a), acknowledging that keymapper users (be they technophiles, power users, people with special needs or the large share of games) are likely to enjoy the slick, more powerful UI. If you prefer to play safe and strictly minimize support requests, (c) is the obvious choice. Lastly, (b) could be worthwhile if combined with an opt-in setting "show secondary sorting chips" that behaves like (c) if unset.

sds100 commented:
Thanks @GfEW for the summary, I agree. Considering most of the work has been done already, and someone will probably request it at some point, I would like to go with option A (vertical list). @maksimowiczm is this alright?

GfEW commented:
Laggard thought on a fourth variant that might be worth considering before going final:

#### (d) fixed chips (multi sort, temporal order)
Keep chips fixed like (c), but maintain previously tapped sort criteria subordinary to the newly tapped one, with declining priority. This follows the idea of ageing briefly discussed in (b) above: "the chip tapped last matters most".

Technically, each of the four priority levels 1/2/3/4 is always(*) bound to one of the four chips, and whichever chip is tapped, becomes the new 1 (unless it's 1 already, in which case its sort direction is reversed), whilst 2/3/4 keep their old relative order.

* `+` sorting by (fixed) chips is a familiar android UI paradigm 
* `+` makes _simple sort_ as simple as with (c)
* `+` allows _multi sort_ nearly(**) as powerful as with (a)
* `*` multi sort priorisation based on recency intuitively blends in with simple sort, uninterested users won't notice
* `*` the arrangement of chips doesn't matter (unlike with (b)), this might help with future additions of e. g. filters
* `-` proper visualization is challenging - could the chips pale gradually by each demotion, so there are four(***) distinct levels of tint, from _"full"_ to _"none"_?

I'm not saying (d) was strictly better than (a), but its specific advantages could indeed weigh more.

Over to you.

<details><summary>(*)</summary>Initial order (after installation) can start e. g. 1/2/3/4 from left to right, but this ceases to matter as chips get tapped.

</details>
<details><summary>(**)</summary>Only "nearly" as powerful as (a) because only the main key could be reversed. (Very dedicated users could still work around this limitation by temporarily promoting the to-be-reversed subordinary key to prio 1, then recreating the other prios.)

</details>
<details><summary>(***)</summary>To account for high contrast themes, b/w E-ink etc. that might reduce the four shades to two, "full" tint must stand out clearly, as telling prio 1 apart from 2/3/4 is most essential.

maksimowiczm commented:
I like the tinted chips idea, but I think it will be a pain to implement due to the different themes, colors, and the way the Material Design color system works.

I think the initial idea of the sortable list is the best. The only potential problem I see is that if more sorting options are added, its size might become too large.

---

I noticed that there is a jetpack compose setup, and from a technical point of view, I think it would be a good idea to use compose to implement this feature. Ironically, compose doesn't provide an easy way to implement drag-and-drop functionality, as it requires a lot of boilerplate code. I also found this library https://github.com/Calvin-LL/Reorderable, which allows for straightforward implementation. The question is, what do you think about adding another dependency?

sds100 commented:
@maksimowiczm Yeah! I'd prefer it if you use Compose for new UI features! And lets stick to the original - maybe we don't have the "options" sort field to give room to grow because there aren't that many options to sort by?

I like to avoid 3rd party libraries because I've already been shot in the foot a couple times in key mapper when they go unmaintained. The years end up flying by 😅 I had to implement drag and drop in Compose and I found an official example that seems to work well. Here is the code...

[DragAndDrop.kt.zip](https://github.com/user-attachments/files/18228183/DragAndDrop.kt.zip)

maksimowiczm commented:
The UI is mostly done. I used some of the code you shared, but I'm not sure if it will work because it bugs a bit on the emulator, it works on my device, though 😆. It definitely needs more testing.

I simplified use cases since there are only 4 sort fields I just dumped them into data store as json. It also avoids unnecessary key map list resorting.

---

Regarding comparators, I think they should be stored in the sorting package. However, I feel there might be a better way to handle this since the constraints and trigger comparators depend on the `Comparable` implementation of `Constraint` and `Trigger` classes. I'm not sure if there's a cleaner alternative to `Comparable` though.

---

Quick demo

[demo.webm](https://github.com/user-attachments/assets/477a95e4-5563-4492-8479-a46fb2e9a172)

I noticed now that cancel didn't work, will fix that later.

GfEW commented:
Are sort settings preserved across restarts?

maksimowiczm commented:
I like the idea of row labels, but I had to attach them to the moving parts because otherwise it would break either accessibility or animations.
I added a 'Reset' button which resets orders (not priority) and moved the 'Apply' button to the bottom because I kept accidentally clicking the reset button instead of apply.
I also added a small help message for users, though I'm not entirely sure about its message.

All sort settings are preserved between restarts.

https://github.com/user-attachments/assets/54a7ff08-00da-4a1f-bb97-688e862c89fa

sds100 commented:
This UI looks good! Can you round the corners of the bottom sheet like I did for the others?

maksimowiczm commented:
I think it is ready now.

GfEW commented:
@maksimowiczm 
> I like the idea of row labels, but I had to attach them to the moving parts

I see. From a pure UI design perspective, moving priority labels along with the movable fields during the animation "as if they were characteristic of those fields" feels wrong. Their entire point is "each priority has its _fixed_ place, no matter which field you move there". Some "edge case" users who'd barely understand the whole multisort thing with fixed, most self-explanatory labels, might get confused due to this. Then again, I guess that's one of those compromises we often need to make at the end. ;-)

That said, you've created a very capable solution!
Congrats you got this working so nicely!

GfEW commented:
On the help message, maybe sth like this?
(Choose version according to available space - you know better how much there is.)

#### short version
```
Drag: adjust priorities
Tap: reverse direction
```

#### medium version
```
Drag items to adjust priorities, tap them to reverse sort direction.
Example: To sort bindings by actions, move "action" to the top."
```

#### long version
```
Drag items to adjust priorities. Top item matters most. You may also tap any item to reverse its sort direction.
Example: To sort bindings mainly by their actions, and secondly by their triggers, move "actions" to first place, and "trigger" second.
```

Please adapt the wording as you see fit, also correcting any English language glitches, as my proficiency is somewhat limited there (I'm not a native speaker).

maksimowiczm commented:
I feel that the help message shouldn't depend on screen size, but it also shouldn't take up the whole screen. I think an elegant solution would be to give the user an option to hide it.

Additionally, an show example button might be helpful.

https://github.com/user-attachments/assets/67225fcd-4713-4288-b41d-e54abb7ab7e0

GfEW commented:
> I feel that the help message shouldn't depend on screen size

Yeah, I didn't mean to adapt text length to the screen size of the particular end device but to choose one variant, based on your general assessment of best fit.

The help message in your demo certainly doesn't err on the "too terse" side. I'm not sure if @sds100 agrees with this degree of verbosity for an auxiliary feature whilst keymapper's core features provide little (if any) inline help. 

That discrepancy of style aside, I think the comprehensive, hidable message, in conjunction with the demo example(*), will do the job of familiarizing inexperienced users with the sort UI, very well.

<details><summary>(*)</summary>The explanation of the example mentions, but then doesn't elaborate on, sort directions, so it depends on the example to fully make sense.

</details>

sds100 commented:
I like the collapsible help message!

sds100 commented on app/src/main/res/drawable/arrow_downward_24.xml:
Do we need this file?

sds100 commented on app/src/main/res/drawable/ic_arrow_upward_24.xml:
Do we need this file?

sds100 commented on app/src/main/java/io/github/sds100/keymapper/sorting/ObserveKeyMapsSorterUseCase.kt:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Observes the key maps comparator based on the user's preferences.
+ */
+class ObserveKeyMapsSorterUseCase(
```
Can we merge all the sorting use cases into one use case? Just call it `SortKeyMapsUseCase`?

sds100 commented:
Sorry @maksimowiczm did you already see my review comments before I submitted them?

maksimowiczm commented:
First time seeing those. I will check them right now.

maksimowiczm commented on app/src/main/res/menu/menu_home.xml:
```diff
@@ -7,4 +7,16 @@
+        android:title="@string/action_sort"
+        app:showAsAction="ifRoom" />
+
+    <item
```
Yep, my bad. It must have slipped out while merging with master.

maksimowiczm commented on app/src/main/java/io/github/sds100/keymapper/sorting/ObserveKeyMapFieldSortOrderUseCase.kt:
```diff
@@ -0,0 +1,37 @@
+ * Observes the sorting order of a key map field, indicating the order in which the specified field
+ * should be sorted.
+ */
+class ObserveKeyMapFieldSortOrderUseCase(
```
I feel you, especially with a mess that was there before.

maksimowiczm commented on app/src/main/java/io/github/sds100/keymapper/data/Keys.kt:
```diff
@@ -68,4 +68,11 @@ object Keys {
     val neverShowDndError = booleanPreferencesKey("key_never_show_dnd_error")
+
+    // true = ascending, false = descending
+    val sortTriggerAscending = booleanPreferencesKey("key_sort_trigger_ascending")
```
This is no longer relevant since I dump everything in json.

maksimowiczm commented on app/src/main/java/io/github/sds100/keymapper/mappings/keymaps/ListKeyMapsUseCase.kt:
```diff
@@ -17,23 +19,25 @@ class ListKeyMapsUseCaseImpl(
+            keyMapEntitiesState.mapData { keyMapEntities ->
+                keyMapEntities
+                    .map { KeyMapEntityMapper.fromEntity(it) }
+                    .sortedWith(sorter)
```
I am not sure if I am missing something.

I can’t just reverse the whole list because it wouldn’t make sense with the field priorities. Instead, I need to reverse specific comparators to get the sorting right.

---

I don't know how to move sorting in a clean way because listing depends entirely on sorting.

Currently it works like this:
1. User sets order 
2. New sorter is created and notifies observers (`ListKeyMapsUseCase`)
3. `ListKeyMapsUseCase` updates the list

If I create a method that takes a list and returns sorted lists, there’s no way for the UI to update itself.

sds100 replied:
Ahh, yes. Lets keep the reverse field then. Can you append the comment above the `reversed` field with what you said so we don't forget in the future.

sds100 replied:
Did it myself so we can wrap this up quickly :)

maksimowiczm commented:
@GfEW I just realized what you meant about these help messages being adjusted according to the available space. I tried to make it work, but it didn’t look right, and I can’t spend infinite time on it, so it is what it is and its cool.

sds100 commented:
This looks good now, I'm just doing some minor refactor/style changes to make it more consistent with the other code then I think its done!

GfEW commented:
> @GfEW I just realized what you meant about these help messages being adjusted according to the available space. I tried to make it work, but it didn’t look right, and I can’t spend infinite time on it, so it is what it is and its cool.

Sure, go ahead. Things can be revisited anyway later on, should the need arise.

maksimowiczm commented on app/src/main/java/io/github/sds100/keymapper/constraints/Constraint.kt:
```diff
@@ -15,20 +15,42 @@ import java.util.UUID
+sealed class Constraint : Comparable<Constraint> {
     val uid: String = UUID.randomUUID().toString()
 
+    override fun compareTo(other: Constraint) = this.javaClass.name.compareTo(other.javaClass.name)
```
I just realized that this might be bad idea since r8 obfuscate class names.

sds100 replied:
Good spot! Can't you just use the Kotlin equality operator `==`?

maksimowiczm replied:
No, because `Comparable` is about ordering, not equality

sds100 replied:
Yeah, just realised! Shouldn't we also be ordering by the action text and not the serialized data?

sds100 replied:
And can't we just use the existing methods for the constraint text? We don't need to make a generic function?

maksimowiczm replied:
I think the best way to sort is by using localized strings to avoid confusion. There are already helper classes available for this use case. 

It can be done for both actions and constrains. I think options and triggers can remain as they are for now.

Not sure if there is anything that could help be sort constraint classes, so I added a helper method `ConstraintUiHelper.getGenericTitle`.

 Let me know what you think.

sds100 replied:
Yeah, we should be using the localized strings. Why do we need `getGenericTitle`? Can't we just use `getTitle`?

Now I really think the sorting should be done on the ListItems, it seems a bit messy to be sorting the data by classes that are used for the list items, then converting them to list items and using those same methods again.

sds100 commented on app/src/main/java/io/github/sds100/keymapper/sorting/comparators/KeyMapConstraintsComparator.kt:
```diff
@@ -0,0 +1,77 @@
+): Int {
+    // If constraints are different, compare their generic titles.
+    //
+    // This ensures that there won't be a case like this:
```
@maksimowiczm I will work on finishing this PR. Why is the first sorting not what we want? I think we need to do it anyway, because in other languages the app name may not necessarily be at the beginning so it will then be sorted by "in foreground" etc.

maksimowiczm replied:
This was the requirement from https://github.com/keymapperorg/KeyMapper/issues/1223.

> Mappings with the same category of constraints (if any) would then be listed one after another, e. g. "app in forground" (if any), then "app not in forground" (if any), then "app playing media" (if any), and so forth.

sds100 replied:
Ah, i better understand, thank you!

sds100 commented:
@maksimowiczm @GfEW I'm now sorting actions and constraints primarily by their type and then by secondary data such as the app name etc. Can you verify it is working as intended?

maksimowiczm commented:
Sorting by types is not alphabetical, but I think it is good enough!
